packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -856,3 +856,68 @@ export function forceRehydration(vm: VM) {
         scheduleRehydration(vm);
     }
 }
+
+export function runFormAssociatedCustomElementCallback(vm: VM, faceCb: () => void) {
+    if (!process.env.IS_BROWSER) {
+        return;
+    }
+
+    const {
+        renderMode,
+        shadowMode,
+        def: { formAssociated },
+    } = vm;
+
+    if (shadowMode === ShadowMode.Synthetic && renderMode !== RenderMode.Light) {
+        throw new Error(
+            'Form associated lifecycle methods are not available in synthetic shadow. Please use native shadow or light DOM.'
+        );
+    }
+
+    // Technically the UpgradableConstructor always sets `static formAssociated` to true but we need to verify that any LWCs
+    // using FACE callbacks also set this value. This is because we plan to remove the UpgradableConstructor in the future and
+    // invoke the LWC constructor directly.
+    if (isUndefined(formAssociated)) {
+        throw new Error(
+            `Form associated lifecycle methods must have the 'static formAssociated' value set in the component's prototype chain.`
```

Comment:
@nolanlawson do you think this should be a dev only error instead?

Reply:
Updated in the latest [commit](https://github.com/salesforce/lwc/pull/3695/commits/2a1fe360257411ee991779bdd7dc2d1bc5a05efc)!